packages/markers/src/browser/problem/problem-widget.tsx

```diff
@@ -158,3 +173,70 @@ export class ProblemMarkerRemoveButton extends React.Component<{ model: ProblemT
         e.stopPropagation();
     }
 }
+
+export class ProblemMarkerCopyMessageButton extends React.Component<{ model: ProblemTreeModel, node: MarkerNode }> {
+
+    render(): React.ReactNode {
+        return <span className='copy-message' title='Copy Message' onClick={this.copyMessage}></span>;
+    }
+
+    protected readonly copyMessage = (e: React.MouseEvent<HTMLElement>) => this.doCopyMessage(e);
+
+    protected doCopyMessage(e: React.MouseEvent<HTMLElement>) {
+        if (ProblemMarker.is(this.props.node.marker)) {
+            const problemMarker = this.props.node.marker;
+            const cpElement = document.createElement('textarea');
+            cpElement.value = problemMarker.data.message;
+            document.body.appendChild(cpElement);
+            cpElement.select();
+            document.execCommand('copy');
+            document.body.removeChild(cpElement);
+        }
+    }
+}
+
+export class ProblemMarkerCopyAllButton extends React.Component<{ model: ProblemTreeModel, node: MarkerNode }> {
+
+    render(): React.ReactNode {
+        return <span className='copy-all' title='Copy' onClick={this.copyAll}></span>;
+    }
+
+    protected readonly copyAll = (e: React.MouseEvent<HTMLElement>) => this.doCopyAll(e);
+
+    protected doCopyAll(e: React.MouseEvent<HTMLElement>) {
+        if (ProblemMarker.is(this.props.node.marker)) {
+            const problemMarker = this.props.node.marker;
+            const cpElement = document.createElement('textarea');
+            cpElement.value = this.serializeProblem(problemMarker);
+            document.body.appendChild(cpElement);
+            cpElement.select();
+            document.execCommand('copy');
+            document.body.removeChild(cpElement);
+        }
+    }
+
+    protected serializeProblem(problemMarker: ProblemMarker): string {
+        const resource = problemMarker.uri;
+        const owner = problemMarker.owner;
+        const code = problemMarker.data.code;
+        const severity = problemMarker.data.severity;
+        const message = problemMarker.data.message;
+        const source = problemMarker.data.source;
+        const startLineNumber = problemMarker.data.range.start.line;
+        const startColumn = problemMarker.data.range.start.character;
+        const endLineNumber = problemMarker.data.range.end.line;
+        const endColumn = problemMarker.data.range.end.character;
+        return JSON.stringify({
```

Comment:
You should return an object where you define the fields directly, instead of assigning a variable for each.